### PR TITLE
Beam-2384 CEIL, FLOOR, TRUNCATE, PI, ATAN2 math functions

### PR DESCRIPTION
**Pull request by app-tarush**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by app-tarush** [comment]
@xumingming Please review this PR #3525.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java [hunk 1]
```diff
@@ -278,6 +283,15 @@ static BeamSqlExpression buildExpression(RexNode rexNode) {
         case "SIGN":
           ret = new BeamSqlSignExpression(subExps);
           break;
+        case "PI":
+          ret = new BeamSqlPiExpression();
+          break;
+        case "ATAN2":
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: Why it's named `ATAN2`?

Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: This is a standard function mysql ref :

ATAN(Y,X), ATAN2(Y,X)

Returns the arc tangent of the two variables X and Y. It is similar to calculating the arc tangent of Y / X, except that the signs of both arguments are used to determine the quadrant of the result.

```mysql
mysql> SELECT ATAN(-2,2);
        -> -0.78539816339745
mysql> SELECT ATAN2(PI(),0);
        -> 1.5707963267949
```

Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: Cool, don't know that before.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@link BeamSqlMathBinaryExpression} for 'ATAN2' function.
+ */
+public class BeamSqlAtan2Expression extends BeamSqlMathBinaryExpression {
+
+  public BeamSqlAtan2Expression(List<BeamSqlExpression> operands) {
+    super(operands);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java]: let's specify the outputType explicitly.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlCeilExpression.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@code BeamSqlMathUnaryExpression} for 'CEIL' function.
+ */
+public class BeamSqlCeilExpression extends BeamSqlMathUnaryExpression {
+
+  public BeamSqlCeilExpression(List<BeamSqlExpression> operands) {
+    super(operands);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlCeilExpression.java]: specify outputType

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlFloorExpression.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@code BeamSqlMathUnaryExpression} for 'FLOOR' function.
+ */
+public class BeamSqlFloorExpression extends BeamSqlMathUnaryExpression {
+
+  public BeamSqlFloorExpression(List<BeamSqlExpression> operands) {
+    super(operands);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlFloorExpression.java]: specify outputType

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPiExpression.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  public BeamSqlPiExpression() {
+    this.outputType = SqlTypeName.DOUBLE;
+  }
+
+  @Override public boolean accept() {
+    return numberOfOperands() == 0;
+  }
+
+  @Override public BeamSqlPrimitive evaluate(BeamSqlRow inputRecord) {
+    return BeamSqlPrimitive.of(SqlTypeName.DOUBLE, Math.PI);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPiExpression.java]: SqlTypeName.DOUBLE -> `this.outputType`

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlTruncateExpression.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@code BeamSqlMathBinaryExpression} for 'TRUNCATE' function.
+ */
+public class BeamSqlTruncateExpression extends BeamSqlMathBinaryExpression {
+
+  public BeamSqlTruncateExpression(List<BeamSqlExpression> operands) {
+    super(operands);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlTruncateExpression.java]: specify outputType

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java (hunk 1 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: 👍

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java (hunk 2 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java]: I will update outputType

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlTruncateExpression.java (hunk 6 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlTruncateExpression.java]: This function outputs different types so this has been initialised to 
```java
SqlTypeName.ANY
```
in the BeamSqlMathBinaryExpression

```java
public abstract class BeamSqlMathBinaryExpression extends BeamSqlExpression {

  public BeamSqlMathBinaryExpression(List<BeamSqlExpression> operands) {
    super(operands, SqlTypeName.ANY);
  }
```

Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlTruncateExpression.java]: having a outputType as `SqlTypeName.ANY` might have issue when writing the integration test, you can take a look at https://github.com/apache/beam/pull/3557 to see how I am handling it.

**Comment by xumingming** [comment]
@takidau can you also take a look at this one?

**Comment by jbonofre** [comment]
R: @jbonofre

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.beam.dsls.sql.interpreter.operator.BeamSqlPrimitive;
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@link BeamSqlMathBinaryExpression} for 'ATAN2' function.
+ */
+public class BeamSqlAtan2Expression extends BeamSqlMathBinaryExpression {
+
+  public BeamSqlAtan2Expression(List<BeamSqlExpression> operands) {
```
Comment (jbonofre) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java]: Maybe specifiying `outputType` in the constructor ?

Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java]: I will update the change

Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlAtan2Expression.java]: let's merge this first if no other pending changes, I'll update the integration test to include these functions.

**Comment by JingsongLi** [comment]
R + @JingsongLi, I'll take a quick look.

**Comment by JingsongLi** [comment]
Merged.

**Comment by XuMingmin** [comment]
Thank you @JingsongLi ! 

@app-tarush feel free to close the PR.

**Comment by app-tarush** [comment]
Thank you @JingsongLi @XuMingmin

**Pull request closed without merging**